### PR TITLE
refactor so that ModelArgs does not depend on tokenizer 

### PR DESCRIPTION
**Pull request by tianyu-l**
Previously we initialize ModelArgs, and then update it dynamically to
1. get `vocab_size` from tokenizer (used for specifying shapes of embedding / output module)
2. get `eos_id` from tokenizer (used for generating block causal attention mask)
3. update `max_seq_len` according to training job `seq_len` (used for precomputing `freqs_cis`)

-------------
1 is causing troubles as we found `vocab_size` for model checkpoints (embedding / output layer) loaded from HF may not be always the same as `tokenizer.get_vocab_size()`. In fact, as long as `vocab_size` in embedding / output layer is larger than tokenizer's `vocab_size`, the training is still OK.

In addition, there have been requests to not let `ModeArgs` and model init depend on tokenizers, so this PR removes 2 and instead send `eos_id` as input to model.

----------------

For 3, there is a caveat that when torchtitan is used for continuing training from a checkpoint, users should be aware that the original model has an intrinsic limit on `max_seq_len`. E.g. for llama 3 it's 8k, for llama 4 Scout it's 1M. Currently torchtitan users could break this limit by specifying an arbitrarily large `--training.seq_len`, whether intentionally or not.

This PR keeps this flexibility for two reasons:
1. when `seq_len` is less than intrinsic `max_seq_len`, we only need to generate `freqs_cis` of `seq_len` both because of resource consideration and because of CP compatibility.
2. when users intentionally want to test long context training / extend to longer context capability, they could still do that.

Instead, this PR adds a warning when getting a `seq_len` larger than the original `max_seq_len`

I noticed that llama "official" implementation also allows this https://github.com/meta-llama/llama-models/blob/main/models/llama4/generation.py#L72

### CODE REVIEW

**Comment by lkhphuc** [comment]
> In addition, there have been requests to not let ModeArgs and model init depend on tokenizers, so this PR removes 2 and instead send eos_id as input to model.

Hi, just want to put my case here for your consideration.
I don't know about the arguments to remove the tokenizer from the model_args's `update_from_config`, but in our uses case (multimodal training) we have about ~10 special tokens (image/video/diffusion start/end tokens)  that is being set in this function.
Modifying train.py to pass all these additional tokens in forward and flex attention init seems less ideal than adding these tokens to our model-specific ModelArgs, as we are trying to keep our train.py to be as close as possible to upstream.

In torchtitan's main with only one `eos_id` I guess it does not matter, but I would argue passing `tokenizer` to `update_from_config` would be more fork-friendly.

**Comment by tianyu-l** [comment]
@lkhphuc 
Thanks for the feedback!

Do you have your own train.py or do you reuse train.py?
If we change the interface to allow optional kwargs (e.g. tokenizer) but don't use it in our train.py, would it be helpful?

**Comment by lkhphuc** [comment]
> Do you have your own train.py or do you reuse train.py?
 
 I modify the upstream train.py directly and try to keep the changes to be minimal and in an additive manner. This ways I can quickly pull upstream changes. But of course sometimes conflict still happens but I think that's the price worth paying for to keep up to date.
 
 >  If we change the interface to allow optional kwargs (e.g. tokenizer) but don't use it in our train.py, would it be helpful?

Yes thanks. this would be good compromise, as then I can just pass in the tokenizer in one place in train.py and keep the same behaviour as currently is.

**Comment by tianyu-l** [comment]
@lkhphuc please see if the changes now look OK to you

The direct reason we are factoring tokenizer `eos_id` dependency out of `ModelArgs` is because
We would like torchtitan to be useful as a post-training component, where downstream libraries can take dependency on a truncated version of torchtitan Trainer. Very often the downstream library would like to have their own tokenizer and dataloader, which we don't have access to during model initialization.

File: torchtitan/models/deepseek_v3/model/args.py [hunk 1]
```diff
@@ -92,14 +92,25 @@ class DeepSeekV3ModelArgs(BaseModelArgs):
...
-        self.max_seq_len = job_config.training.seq_len
+    def update_from_config(self, job_config: JobConfig, **kwargs) -> None:
+        seq_len = job_config.training.seq_len
+        if seq_len > self.max_seq_len:
+            logger.warning(
+                f"Sequence length {seq_len} exceeds original maximum {self.max_seq_len}."
+            )
+        self.max_seq_len = seq_len
+
+        if self.use_grouped_mm and not has_cuda_capability(9, 0):
```
Comment (wwwjn) [inline torchtitan/models/deepseek_v3/model/args.py:104]: Thank you! This is a nice change and I always want to add this :)

**Review by wwwjn — APPROVED** [review]
LGTM, very nice changes! CI is working and you could retry it now
